impl/src/ref_conv.rs

```diff
@@ -0,0 +1,312 @@
+use crate::utils::{add_where_clauses_for_new_ident, Either};
+use proc_macro2::{Ident, TokenStream};
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream, Result},
+    spanned::Spanned,
+    DeriveInput, Field, Fields, Index, Token,
+};
+
+pub fn expand(
+    input: &DeriveInput,
+    trait_ident: &Ident,
+    method_ident: &Ident,
+    conv_type: &Ident,
+    mutability: Option<&Token![mut]>,
+) -> syn::Result<TokenStream> {
+    let trait_path = quote! { ::derive_more::#trait_ident };
+
+    let field_args = extract_field_args(input, trait_ident, method_ident)?;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+    let input_type = &input.ident;
+
+    let sub_items: Vec<_> = field_args
+        .into_iter()
+        .map(
+            |FieldArgs {
+                 forward,
+                 field,
+                 ident,
+             }| {
+                let member = quote! { self.#ident };
+                let field_type = &field.ty;
+                if forward {
+                    let trait_path = quote! { #trait_path<#conv_type> };
+                    let type_where_clauses = quote! {
+                        where #field_type: #trait_path
+                    };
+                    let new_generics = add_where_clauses_for_new_ident(
+                        &input.generics,
+                        &[field],
+                        conv_type,
+                        type_where_clauses,
+                        false,
+                    );
+                    let (impl_generics, _, where_clause) =
+                        new_generics.split_for_impl();
+                    let casted_trait = quote! { <#field_type as #trait_path> };
+                    (
+                        quote! { #casted_trait::#method_ident(& #mutability #member) },
+                        quote! { #impl_generics },
+                        quote! { #where_clause },
+                        quote! { #trait_path },
+                        quote! { & #mutability #conv_type },
+                    )
+                } else {
+                    (
+                        quote! { & #mutability #member },
+                        quote! { #impl_generics },
+                        quote! { #where_clause },
+                        quote! { #trait_path<#field_type> },
+                        quote! { & #mutability #field_type },
+                    )
+                }
+            },
+        )
+        .collect();
+    let bodies = sub_items.iter().map(|i| &i.0);
+    let impl_generics = sub_items.iter().map(|i| &i.1);
+    let where_clauses = sub_items.iter().map(|i| &i.2);
+    let trait_paths = sub_items.iter().map(|i| &i.3);
+    let return_types = sub_items.iter().map(|i| &i.4);
+
+    Ok(quote! {#(
+        #[automatically_derived]
+        impl #impl_generics #trait_paths for #input_type #ty_generics #where_clauses {
```

Comment:
@MegaBluejay I think, here it would be much better if instead of making multiple iterators for each part to use in final `quote! {}`, we just make an iterator of final `quote! {}`s, so we don't squat with those iterators redudantly.

```rust
field_args.into_iter().map(|....| {
    quote! {
        #[automatically_derived]
        impl #impl_generics #trait_paths for #input_type #ty_generics #where_clauses {
            #[inline]
            fn #method_ident(& #mutability self) -> #return_types {
                #bodies
            }
        }
    }
}).collect()
```